File: go/output.go
package gencmu

import (
	"sort"
	"strings"
)

// resultFormat is the version of docs/output.md.
const resultFormat = 1

// MarshalResult writes the canonical JSON of a result (docs/output.md).
func MarshalResult(result *ParseResult) ([]byte, error) {
	var w jsonWriter
	w.raw(`{"format":`)
	w.int(resultFormat)
	w.raw(`,"ok":`)
	w.bool(result.OK)
	w.raw(`,"stages":[`)
	for i := range result.Stages {
		if i > 0 {
			w.raw(",")
		}
		writeStage(&w, &result.Stages[i])
	}
	w.raw(`],"tree":`)
	writeNode(&w, result.Tree)
	w.raw(`,"error":`)
	writeError(&w, result.Error)
	w.raw("}")
	return w.buf, nil
}

func writeStage(w *jsonWriter, s *Stage) {
	w.raw(`{"name":`)
	w.str(s.Name)
	w.raw(`,"verdict":`)
	if s.Verdict == "" {
		w.raw("null")
	} else {
		w.str(s.Verdict)
	}
	if s.Verdict == VerdictTie {
		w.raw(`,"witness":[`)
		for i, a := range s.Witness {
			if i > 0 {
				w.raw(",")
			}
			writeAction(w, a)
		}
		w.raw(`],"tied":`)
		writeNode(w, s.Tied)
	}
	if s.Output != nil {
		w.raw(`,"output":[`)
		for i := range s.Output {
			if i > 0 {
				w.raw(",")
			}
			writeToken(w, &s.Output[i])
		}
		w.raw("]")
	}
	w.raw("}")
}

func writeTags(w *jsonWriter, tags map[string]bool) {
	names := make([]string, 0, len(tags))
	for n := range tags {
		names = append(names, n)
	}
	sort.Strings(names)
	w.raw("{")
	for i, n := range names {
		w.key(i == 0, n)
		w.bool(tags[n])
	}
	w.raw("}")
}

func writeToken(w *jsonWriter, t *Token) {
	w.raw(`{"text":`)
	w.str(t.Text)
	w.raw(`,"phonemes":`)
	w.str(t.Phonemes)
	w.raw(`,"tags":`)
	writeTags(w, t.Tags)
	w.raw(`,"span":`)
	w.pair(t.Span)
	w.raw(`,"source":`)
	w.pair(t.Source)
	if t.InsertedBy != "" {
		w.raw(`,"insertedBy":`)
		w.str(t.InsertedBy)
	}
	w.raw("}")
}

func writeAction(w *jsonWriter, a Action) {
	switch {
	case a.Read != nil:
		w.raw(`{"read":{"token":`)
		w.int(a.Read.Token)
		w.raw(`,"terminal":`)
		w.str(a.Read.Terminal)
		w.raw("}}")
	case a.Close != nil:
		w.raw(`{"close":{"rule":`)
		w.str(a.Close.Rule)
		w.raw(`,"production":`)
		w.int(a.Close.Production)
		w.raw(`,"span":`)
		w.pair(a.Close.Span)
		w.raw("}}")
	default:
		w.raw("null")
	}
}

// writeNode writes a tree; trees are as deep as the grammar nests, so this
// keeps its own stack.
func writeNode(w *jsonWriter, root *Node) {
	if root == nil {
		w.raw("null")
		return
	}
	type frame struct {
		n    *Node
		next int
	}
	stack := []frame{{n: root}}
	open := func(n *Node) {
		w.raw(`{"kind":`)
		w.str(n.Kind)
		switch n.Kind {
		case KindRule:
			w.raw(`,"rule":`)
			w.str(n.Rule)
		default:
			w.raw(`,"terminal":`)
			w.str(n.Terminal)
			if n.Kind == KindToken {
				w.raw(`,"token":`)
				w.int(n.Token)
			}
		}
		w.raw(`,"span":`)
		w.pair(n.Span)
		w.raw(`,"source":`)
		w.pair(n.Source)
		if n.Kind == KindRule {
			w.raw(`,"tags":`)
			writeTags(w, n.Tags)
			w.raw(`,"children":[`)
		} else {
			w.raw("}")
		}
	}
	open(root)
	if root.Kind != KindRule {
		return
	}
	for len(stack) > 0 {
		f := &stack[len(stack)-1]
		if f.next == len(f.n.Children) {
			w.raw("]}")
			stack = stack[:len(stack)-1]
			continue
		}
		c := f.n.Children[f.next]
		if f.next > 0 {
			w.raw(",")
		}
		f.next++
		open(c)
		if c.Kind == KindRule {
			stack = append(stack, frame{n: c})
		}
	}
}

func writeError(w *jsonWriter, e *ParseError) {
	if e == nil {
		w.raw("null")
		return
	}
	w.raw(`{"kind":`)
	w.str(e.Kind)
	if e.Stage != "" {
		w.raw(`,"stage":`)
		w.str(e.Stage)
	}
	if e.Token != nil {
		w.raw(`,"token":`)
		w.int(*e.Token)
	}
	if e.Source != nil {
		w.raw(`,"source":`)
		w.pair(*e.Source)
	}
	if e.Document != "" {
		w.raw(`,"document":`)
		w.str(e.Document)
	}
	if e.Line > 0 {
		w.raw(`,"line":`)
		w.int(e.Line)
		w.raw(`,"column":`)
		w.int(e.Column)
	}
	switch e.Kind {
	case ErrorRejected:
		w.raw(`,"expected":[`)
		for i, x := range e.Expected {
			if i > 0 {
				w.raw(",")
			}
			w.raw(`{"terminal":`)
			w.str(x.Terminal)
			w.raw(`,"rules":[`)
			for j, r := range x.Rules {
				if j > 0 {
					w.raw(",")
				}
				w.str(r)
			}
			w.raw("]}")
		}
		w.raw("]")
	case ErrorAmbiguous:
		w.raw(`,"readings":[`)
		for i, r := range e.Readings {
			if i > 0 {
				w.raw(",")
			}
			writeNode(w, r)
		}
		w.raw("]")
	}
	w.raw(`,"message":`)
	w.str(e.Message)
	w.raw("}")
}

// BracketOptions are the options of the bracket rendering.
type BracketOptions struct {
	// ShowElided shows elided terminators as ⟨ku⟩.
	ShowElided bool
}

// Brackets renders a result's tree as nested groups (docs/output.md,
// "Brackets"): "" for a result without a tree.
func Brackets(result *ParseResult, options BracketOptions) string {
	if result == nil || result.Tree == nil || len(result.Stages) == 0 {
		return ""
	}
	input := result.Stages[len(result.Stages)-1].Input
	// Render bottom-up: each node becomes empty, a label, or a group.
	type rendered struct {
		leaf  string
		group []*rendered
		empty bool
	}
	type frame struct {
		n    *Node
		kids []*rendered
	}
	var result2 *rendered
	stack := []*frame{{n: result.Tree}}
	finish := func(f *frame) *rendered {
		switch f.n.Kind {
		case KindToken:
			t := input[f.n.Token]
			if t.Phonemes != "" {
				return &rendered{leaf: t.Phonemes}
			}
			return &rendered{leaf: t.Text}
		case KindElided:
			if options.ShowElided {
				return &rendered{leaf: "⟨" + strings.ToLower(f.n.Terminal) + "⟩"}
			}
			return &rendered{empty: true}
		}
		var kept []*rendered
		for _, k := range f.kids {
			if !k.empty {
				kept = append(kept, k)
			}
		}
		switch len(kept) {
		case 0:
			return &rendered{empty: true}
		case 1:
			return kept[0]
		}
		return &rendered{group: kept}
	}
	for len(stack) > 0 {
		f := stack[len(stack)-1]
		if f.n.Kind == KindRule && len(f.kids) < len(f.n.Children) {
			stack = append(stack, &frame{n: f.n.Children[len(f.kids)]})
			continue
		}
		r := finish(f)
		stack = stack[:len(stack)-1]
		if len(stack) == 0 {
			result2 = r
		} else {
			parent := stack[len(stack)-1]
			parent.kids = append(parent.kids, r)
		}
	}
	var b strings.Builder
	var write func(r *rendered, depth int)
	write = func(r *rendered, depth int) {
		if r.group == nil {
			b.WriteString(r.leaf)
			return
		}
		open, close := "([{"[depth%3], ")]}"[depth%3]
		b.WriteByte(open)
		for i, k := range r.group {
			if i > 0 {
				b.WriteByte(' ')
			}
			write(k, depth+1)
		}
		b.WriteByte(close)
	}
	if !result2.empty {
		write(result2, 0)
	}
	return b.String()
}

File: lib/rust/src/output.rs
//! The canonical JSON of a result and the bracket rendering
//! (`docs/output.md`), written by hand and without recursion.

use std::ops::Range;

use crate::json::write_str;
use crate::result::{Action, Node, NodeKind, ParseError, ParseErrorKind, ParseResult, Stage, Tags, Token, Verdict};

fn write_range(out: &mut String, range: &Range<usize>) {
    out.push('[');
    out.push_str(&range.start.to_string());
    out.push(',');
    out.push_str(&range.end.to_string());
    out.push(']');
}

fn write_tags(out: &mut String, tags: &Tags) {
    out.push('{');
    for (index, (name, strong)) in tags.iter().enumerate() {
        if index > 0 {
            out.push(',');
        }
        write_str(out, name);
        out.push_str(if *strong { ":true" } else { ":false" });
    }
    out.push('}');
}

fn write_token(out: &mut String, token: &Token) {
    out.push_str("{\"text\":");
    write_str(out, &token.text);
    out.push_str(",\"phonemes\":");
    write_str(out, token.phonemes.as_deref().unwrap_or(""));
    out.push_str(",\"tags\":");
    write_tags(out, &token.tags);
    out.push_str(",\"span\":");
    write_range(out, &token.span);
    out.push_str(",\"source\":");
    write_range(out, &token.source);
    if let Some(rule) = &token.inserted_by {
        out.push_str(",\"insertedBy\":");
        write_str(out, rule);
    }
    out.push('}');
}

enum Work<'a> {
    Node(&'a Node),
    Text(&'static str),
}

/// Writes a node as canonical JSON.
pub(crate) fn write_node(out: &mut String, root: &Node) {
    let mut stack = vec![Work::Node(root)];
    while let Some(work) = stack.pop() {
        let node = match work {
            Work::Text(text) => {
                out.push_str(text);
                continue;
            }
            Work::Node(node) => node,
        };
        match node.kind {
            NodeKind::Rule => {
                out.push_str("{\"kind\":\"rule\",\"rule\":");
                write_str(out, node.rule.as_deref().unwrap_or(""));
                out.push_str(",\"span\":");
                write_range(out, &node.span);
                out.push_str(",\"source\":");
                write_range(out, &node.source);
                out.push_str(",\"tags\":");
                write_tags(out, &node.tags);
                out.push_str(",\"children\":[");
                stack.push(Work::Text("]}"));
                for (index, child) in node.children.iter().enumerate().rev() {
                    stack.push(Work::Node(child));
                    if index > 0 {
                        stack.push(Work::Text(","));
                    }
                }
            }
            NodeKind::Token => {
                out.push_str("{\"kind\":\"token\",\"terminal\":");
                write_str(out, node.terminal.as_deref().unwrap_or(""));
                out.push_str(",\"token\":");
                out.push_str(&node.token.unwrap_or(0).to_string());
                out.push_str(",\"span\":");
                write_range(out, &node.span);
                out.push_str(",\"source\":");
                write_range(out, &node.source);
                out.push('}');
            }
            NodeKind::Elided => {
                out.push_str("{\"kind\":\"elided\",\"terminal\":");
                write_str(out, node.terminal.as_deref().unwrap_or(""));
                out.push_str(",\"span\":");
                write_range(out, &node.span);
                out.push_str(",\"source\":");
                write_range(out, &node.source);
                out.push('}');
            }
        }
    }
}

fn write_action(out: &mut String, action: &Action) {
    match action {
        Action::Read { token, terminal } => {
            out.push_str("{\"read\":{\"token\":");
            out.push_str(&token.to_string());
            out.push_str(",\"terminal\":");
            write_str(out, terminal);
            out.push_str("}}");
        }
        Action::Close { rule, production, span } => {
            out.push_str("{\"close\":{\"rule\":");
            write_str(out, rule);
            out.push_str(",\"production\":");
            out.push_str(&production.to_string());
            out.push_str(",\"span\":");
            write_range(out, span);
            out.push_str("}}");
        }
    }
}

fn write_stage(out: &mut String, stage: &Stage) {
    out.push_str("{\"name\":");
    write_str(out, &stage.name);
    out.push_str(",\"verdict\":");
    out.push_str(match stage.verdict {
        None => "null",
        Some(Verdict::Unique) => "\"unique\"",
        Some(Verdict::Resolved) => "\"resolved\"",
        Some(Verdict::Tie) => "\"tie\"",
    });
    if let Some([first, second]) = &stage.witness {
        out.push_str(",\"witness\":[");
        write_action(out, first);
        out.push(',');
        write_action(out, second);
        out.push(']');
    }
    if let Some(tied) = &stage.tied {
        out.push_str(",\"tied\":");
        write_node(out, tied);
    }
    if let Some(output) = &stage.output {
        out.push_str(",\"output\":[");
        for (index, token) in output.iter().enumerate() {
            if index > 0 {
                out.push(',');
            }
            write_token(out, token);
        }
        out.push(']');
    }
    out.push('}');
}

fn write_error(out: &mut String, error: &ParseError) {
    out.push_str("{\"kind\":");
    out.push_str(match error.kind {
        ParseErrorKind::Rejected => "\"rejected\"",
        ParseErrorKind::Ambiguous => "\"ambiguous\"",
        ParseErrorKind::Grammar => "\"grammar\"",
    });
    if let Some(stage) = &error.stage {
        out.push_str(",\"stage\":");
        write_str(out, stage);
    }
    if let Some(token) = error.token {
        out.push_str(",\"token\":");
        out.push_str(&token.to_string());
    }
    if let Some(source) = &error.source {
        out.push_str(",\"source\":");
        write_range(out, source);
    }
    if let Some(document) = &error.document {
        out.push_str(",\"document\":");
        write_str(out, document);
    }
    if let Some(line) = error.line {
        out.push_str(",\"line\":");
        out.push_str(&line.to_string());
    }
    if let Some(column) = error.column {
        out.push_str(",\"column\":");
        out.push_str(&column.to_string());
    }
    match error.kind {
        ParseErrorKind::Rejected => {
            out.push_str(",\"expected\":[");
            for (index, expected) in error.expected.iter().enumerate() {
                if index > 0 {
                    out.push(',');
                }
                out.push_str("{\"terminal\":");
                write_str(out, &expected.terminal);
                out.push_str(",\"rules\":[");
                for (index, rule) in expected.rules.iter().enumerate() {
                    if index > 0 {
                        out.push(',');
                    }
                    write_str(out, rule);
                }
                out.push_str("]}");
            }
            out.push(']');
        }
        ParseErrorKind::Ambiguous => {
            out.push_str(",\"readings\":[");
            for (index, reading) in error.readings.iter().enumerate() {
                if index > 0 {
                    out.push(',');
                }
                write_node(out, reading);
            }
            out.push(']');
        }
        ParseErrorKind::Grammar => {}
    }
    out.push_str(",\"message\":");
    write_str(out, &error.message);
    out.push('}');
}

/// Writes a result as the canonical JSON of `docs/output.md`: keys in the
/// documented order, no whitespace, non-ASCII characters as themselves.
pub fn to_json(result: &ParseResult) -> String {
    let mut out = String::new();
    out.push_str("{\"format\":1,\"ok\":");
    out.push_str(if result.ok { "true" } else { "false" });
    out.push_str(",\"stages\":[");
    for (index, stage) in result.stages.iter().enumerate() {
        if index > 0 {
            out.push(',');
        }
        write_stage(&mut out, stage);
    }
    out.push_str("],\"tree\":");
    match &result.tree {
        Some(tree) => write_node(&mut out, tree),
        None => out.push_str("null"),
    }
    out.push_str(",\"error\":");
    match &result.error {
        Some(error) => write_error(&mut out, error),
        None => out.push_str("null"),
    }
    out.push('}');
    out
}

/// Writes a node as canonical JSON (`docs/output.md`, "Node").
pub fn node_to_json(node: &Node) -> String {
    let mut out = String::new();
    write_node(&mut out, node);
    out
}

/// A rendered node: empty, a label, or a group of rendered members.
enum Rendered {
    Empty,
    Leaf(String),
    Group(Vec<usize>),
}

/// Renders a tree as brackets (`docs/output.md`, "Brackets"), labelling
/// token nodes from `tokens`, the input of the tree's stage.
pub(crate) fn brackets(tree: &Node, tokens: &[Token], show_elided: bool) -> String {
    // Bottom-up: render every node into an arena, collapsing as we go.
    let mut order: Vec<(&Node, Option<usize>)> = vec![(tree, None)];
    let mut index = 0;
    while index < order.len() {
        let node = order[index].0;
        for child in &node.children {
            order.push((child, Some(index)));
        }
        index += 1;
    }
    let mut arena: Vec<Rendered> = Vec::with_capacity(order.len());
    let mut members: Vec<Vec<usize>> = vec![Vec::new(); order.len()];
    for _ in 0..order.len() {
        arena.push(Rendered::Empty);
    }
    for index in (0..order.len()).rev() {
        let (node, parent) = order[index];
        let rendered = match node.kind {
            NodeKind::Token => {
                let token = node.token.and_then(|token| tokens.get(token));
                let label = match token {
                    Some(token) => match token.phonemes.as_deref() {
                        // Each pause is written as a space.
                        Some(phonemes) if !phonemes.is_empty() => phonemes.replace('.', " "),
                        _ => token.text.clone(),
                    },
                    None => String::new(),
                };
                // A token is never an empty node, even when its label is
                // empty, as an empty zoi quotation's is.
                Rendered::Leaf(label)
            }
            NodeKind::Elided => {
                if show_elided {
                    Rendered::Leaf(format!("⟨{}⟩", node.terminal.as_deref().unwrap_or("").to_lowercase()))
                } else {
                    Rendered::Empty
                }
            }
            NodeKind::Rule => {
                let mut kept = std::mem::take(&mut members[index]);
                kept.reverse();
                match kept.len() {
                    0 => Rendered::Empty,
                    1 => {
                        let only = kept[0];
                        std::mem::replace(&mut arena[only], Rendered::Empty)
                    }
                    _ => Rendered::Group(kept),
                }
            }
        };
        let empty = matches!(rendered, Rendered::Empty);
        arena[index] = rendered;
        if let Some(parent) = parent {
            if !empty {
                members[parent].push(index);
            }
        }
    }
    // Top-down: write groups with brackets by depth.
    enum Write {
        Node(usize, usize),
        Text(&'static str),
    }
    let mut out = String::new();
    let mut stack = vec![Write::Node(0, 0)];
    while let Some(work) = stack.pop() {
        match work {
            Write::Text(text) => out.push_str(text),
            Write::Node(index, depth) => match &arena[index] {
                Rendered::Empty => {}
                Rendered::Leaf(label) => out.push_str(label),
                Rendered::Group(items) => {
                    let (open, close) = match depth % 3 {
                        0 => ("(", ")"),
                        1 => ("[", "]"),
                        _ => ("{", "}"),
                    };
                    out.push_str(open);
                    stack.push(Write::Text(close));
                    for (position, &item) in items.iter().enumerate().rev() {
                        stack.push(Write::Node(item, depth + 1));
                        if position > 0 {
                            stack.push(Write::Text(" "));
                        }
                    }
                }
            },
        }
    }
    out
}

/// Renders a result's tree as brackets (`docs/output.md`), with elided
/// terminators shown as `⟨ku⟩` or hidden; empty when there is no tree.
pub fn to_brackets(result: &ParseResult, show_elided: bool) -> String {
    match (&result.tree, result.stages.last()) {
        (Some(tree), Some(stage)) => brackets(tree, &stage.input, show_elided),
        _ => String::new(),
    }
}
